snuba/datasets/groups.py

```diff
@@ -0,0 +1,119 @@
+    ProjectExtension,
+    QueryExtension,
+)
+from snuba.query.query import AliasedColumn
+from snuba.query.timeseries import TimeSeriesExtension
+
+
+class Groups(TimeSeriesDataset):
+    """
+    Experimental dataset that provides Groups data joined with
+    the events table.
+    """
+
+    EVENTS_ALIAS = "events"
+    GROUPS_ALIAS = "groups"
+
+    ALIASED_COLUMN_REGEX = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z0-9_\.]+)$")
+
+    def __init__(self) -> None:
+        self.__grouped_message = get_dataset("groupedmessage")
+        self.__events = get_dataset("events")
+
+        join_structure = JoinStructure(
+            left_source=SchemaJoinedSource(
+                self.GROUPS_ALIAS,
+                self.__grouped_message.get_dataset_schemas().get_read_schema(),
+            ),
+            right_source=SchemaJoinedSource(
+                self.EVENTS_ALIAS,
+                self.__events.get_dataset_schemas().get_read_schema(),
+            ),
+            mapping=[
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="project_id"),
+                    right=JoinConditionExpression(table_alias="events", column="project_id"),
+                ),
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="id"),
+                    right=JoinConditionExpression(table_alias="events", column="group_id"),
+                ),
+            ],
+            join_type=JoinType.LEFT,
+        )
+
+        schema = JoinedSchema(join_structure)
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            processor=None,
+            default_topic=None,
+            time_group_columns={
+                'events.time': 'events.timestamp',
+            },
+        )
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def __parse_qualified_column(self, column_name: str) -> AliasedColumn:
+        # TODO: This logic should be general and in the Query class.
+        # cannot do that yet since the column processing methods like
+        # column_expr do not have access to the Query yet.
+        match = self.ALIASED_COLUMN_REGEX.match(column_name)
+        if not match or not match[1] in [self.EVENTS_ALIAS, self.GROUPS_ALIAS]:
+            return (None, column_name)
+        else:
+            return (match[1], match[2])
+
+    def column_expr(self, column_name, body):
+        aliased_column = self.__parse_qualified_column(column_name)
+        table_alias = aliased_column[0]
+        if table_alias == self.GROUPS_ALIAS:
+            return self.__grouped_message.column_expr(column_name, body)
+        else:
+            # this allows to delegate columns like tags[blabla] to
```

Comment:
I think the right way to solve this is to change the way we parse the query and split the logic that concerns formatting from the logic that concerns parsing and transformation:
Example:
- right after json validation we build the Query object (as we do today)
- the query object will contain an abstract list of selected columns expressed as an AST of dataclasses parsed from the string (instead of the bare strings we have today).
- dataset can provide some custom dataclasses to use (Tags for example)
- The, query organized this way, goes through the query processors that can edit it.
- The query is passed to the Clickhouse query class and the formatting logic happens there (example transforming a FunctionCall class into function(x). This logic, today is conflated into column_expr with more abstract query editing.

This all anyway, is not needed for this dataset that can delegate column_expr to events.